src/VirtoCommerce.Platform.Assets.AzureBlobStorage/AzureBlobProvider.cs

```diff
@@ -279,58 +282,63 @@ protected virtual async Task MoveAsync(string oldUrl, string newUrl, bool isCopy
+            var blobItems = blobContainer.GetBlobsAsync(prefix: oldPath);
 
-            var items = blobContainer
-                .ListBlobsSegmentedAsync(oldPath, true, BlobListingDetails.All, null, null, null, null).Result;
-
-            foreach (var listBlobItem in items.Results)
+            await foreach (var blobItem in blobItems)
             {
-                var blobName = isFolderRename
-                    ? listBlobItem.Uri.AbsoluteUri
-                    : listBlobItem.StorageUri.PrimaryUri.ToString();
-
-                blobName = blobName.Replace("%20", " ");
-
+                var blobName = UrlHelperExtensions.Combine(containerName, blobItem.Name);
                 var newBlobName = blobName.Replace(oldPath, newPath);
 
-                await MoveBlob(blobContainer, blobName, newBlobName, isCopy);
+                taskList.Add(MoveBlob(blobContainer, blobName, newBlobName, isCopy));
             }
+
+            await Task.WhenAll(taskList);
         }
 
         /// <summary>
-        /// Move blob new url and remove old blob
+        /// Move blob new URL and remove old blob
         /// </summary>
         /// <param name="container"></param>
         /// <param name="oldUrl"></param>
         /// <param name="newUrl"></param>
         /// <param name="isCopy"></param>
-        private async Task MoveBlob(CloudBlobContainer container, string oldUrl, string newUrl, bool isCopy)
+        private async Task MoveBlob(BlobContainerClient container, string oldUrl, string newUrl, bool isCopy)
         {
-            var targetPath = oldUrl.EndsWith(_cloudBlobClient.DefaultDelimiter)
+            var targetPath = newUrl.EndsWith(Delimiter)
                 ? GetDirectoryPathFromUrl(newUrl)
                 : GetFilePathFromUrl(newUrl);
 
-            var target = container.GetBlockBlobReference(targetPath);
-
-            await container.CreateIfNotExistsAsync();
+            var target = container.GetBlockBlobClient(targetPath);
 
             if (!await target.ExistsAsync())
             {
-                var soursePath = oldUrl.EndsWith(_cloudBlobClient.DefaultDelimiter)
+                var soursePath = oldUrl.EndsWith(Delimiter)
                     ? GetDirectoryPathFromUrl(oldUrl)
                     : GetFilePathFromUrl(oldUrl);
 
-                var sourse = container.GetBlockBlobReference(soursePath);
-
-                var sourceExists = await sourse.ExistsAsync();
+                var sourceBlob = container.GetBlockBlobClient(soursePath);
 
-                if (sourceExists)
+                if (await sourceBlob.ExistsAsync())
                 {
-                    await target.StartCopyAsync(sourse);
+                    // Lease the source blob for the copy operation to prevent another client from modifying it.
+                    var lease = sourceBlob.GetBlobLeaseClient();
+                    // Specifying -1 for the lease interval creates an infinite lease.
```

Comment:
If the application crashes during a copy phase a source lease will not be released and continue being in the locked state.
Did you consider this case?

Reply:
Don't see any solution in this case: removing the leasing code